nx/lib/nx.ex

```diff
@@ -6881,6 +6881,216 @@ defmodule Nx do
+
+      iex> Nx.mode(Nx.tensor([[1, 2, 2, 3, 5], [1, 1, 76, 8, 1]]), axis: 0)
+      #Nx.Tensor<
+        s64[5]
+        [1, 1, 2, 3, 1]
+      >
+
+      iex> Nx.mode(Nx.tensor([[1, 2, 2, 3, 5], [1, 1, 76, 8, 1]]), axis: 1)
+      #Nx.Tensor<
+        s64[2]
+        [2, 1]
+      >
+
+      iex> Nx.mode(Nx.tensor([[[[1]]]]), axis: 1)
+      #Nx.Tensor<
+        s64[1][1][1]
+        [
+          [
+            [1]
+          ]
+        ]
+      >
+
+  ### Keeping axis
+
+      iex> Nx.mode(Nx.tensor([[1, 2, 2, 3, 5], [1, 1, 76, 8, 1]]), axis: 1, keep_axis: true)
+      #Nx.Tensor<
+        s64[2][1]
+        [
+          [2],
+          [1]
+        ]
+      >
+
+      iex> Nx.mode(Nx.tensor(1), keep_axis: true)
+      #Nx.Tensor<
+        s64[1]
+        [1]
+      >
+
+      iex> Nx.mode(Nx.tensor([[[1]]]), keep_axis: true)
+      #Nx.Tensor<
+        s64[1]
+        [1]
+      >
+
+      iex> Nx.mode(Nx.tensor([[[[1]]]]), axis: 1, keep_axis: true)
+      #Nx.Tensor<
+        s64[1][1][1][1]
+        [
+          [
+            [
+              [1]
+            ]
+          ]
+        ]
+      >
+  """
+  @doc type: :aggregation, from_backend: false
+  def mode(tensor, opts \\ []) do
+    opts = keyword!(opts, axis: nil, keep_axis: false)
+    %T{shape: shape, names: names} = tensor = to_tensor(tensor)
+
+    axis =
+      if opts[:axis] != nil,
+        do: Nx.Shape.normalize_axis(shape, opts[:axis], names),
+        else: opts[:axis]
+
+    tensor_rank = rank(tensor)
+    tensor_size = size(tensor)
+
+    cond do
+      tensor_rank == 0 ->
+        if opts[:keep_axis], do: new_axis(tensor, -1), else: tensor
+
+      tensor_size == 1 and axis == nil ->
+        tensor = flatten(tensor)
+        res = take(tensor, tensor([0]))
+        if opts[:keep_axis], do: res, else: squeeze(res)
+
+      tensor_size == 1 and axis != nil ->
+        if opts[:keep_axis], do: tensor, else: squeeze(tensor, axes: [axis])
+
+      size(tensor) == 1 ->
+        res = take(tensor, broadcast(0, {tensor_rank}))
+        if opts[:keep_axis], do: res, else: squeeze(res, axes: [axis])
+
+      axis == nil ->
+        tensor = flatten(tensor)
+        mode_1d(tensor, keep_axis: opts[:keep_axis])
+
+      true ->
+        mode_general(tensor, axis: axis, keep_axis: opts[:keep_axis])
+    end
+  end
+
+  defp mode_general(tensor, opts) do
+    tensor_shape = shape(tensor)
+    axis = opts[:axis]
+
+    sorted = sort(tensor, axis: axis)
+
+    size_to_broadcast = tensor_shape |> put_elem(axis, 1)
+
+    group_indices =
+      concatenate(
+        [
+          broadcast(0, size_to_broadcast),
+          not_equal(
+            slice_along_axis(sorted, 0, axis_size(sorted, axis) - 1, axis: axis),
+            slice_along_axis(sorted, 1, axis_size(sorted, axis) - 1, axis: axis)
+          )
+        ],
+        axis: axis
+      )
+      |> cumulative_sum(axis: axis)
+
+    num_elements = tensor_shape |> Tuple.product()
+
+    gen_indices =
```

Comment:
Let's rename to `counting_indices` or something, otherwise it'c easy to confuse for `group_intices`.